OrbitQt/TutorialOverlay.cpp

```diff
@@ -0,0 +1,423 @@
+  ui_->setupUi(this);
+
+  if (parent != nullptr) {
+    setGeometry(QRect(0, 0, parent->rect().width(), parent->rect().height()));
+  }
+  setWindowFlags(Qt::FramelessWindowHint | Qt::WindowSystemMenuHint);
+
+  background_label_ = std::make_unique<QLabel>(this);
+  background_label_->setStyleSheet("background-color: rgba(0, 0, 0, 150);");
+
+  ui_->tabWidget->raise();
+  ui_->titleFrame->raise();
+  ui_->btnClose->raise();
+  QTabBar* tabBar = ui_->tabWidget->findChild<QTabBar*>();
+  tabBar->hide();
+
+  InitAllStepsFromUi();
+}
+
+TutorialOverlay::Step TutorialOverlay::InitializeStep(int tab_index) {
+  QWidget* tab = ui_->tabWidget->widget(tab_index);
+
+  Step step;
+  step.tab_index = tab_index;
+  step.name = tab->objectName().toStdString();
+  step.cutout_widget = tab->findChild<CutoutWidget*>();
+
+  if (step.cutout_widget) {
+    QRect cutout_rect = step.cutout_widget->geometry();
+    for (auto child : tab->findChildren<QWidget*>()) {
+      if (child != step.cutout_widget && child->parent() == tab) {
+        step.hints.push_back(DeriveHintDescription(cutout_rect, child));
+      }
+    }
+  }
+
+  return step;
+}
+
+void TutorialOverlay::InitAllStepsFromUi() {
+  steps_.clear();
+  for (int i = 0; i < ui_->tabWidget->count(); ++i) {
+    const std::string tab_name = ui_->tabWidget->widget(i)->objectName().toStdString();
+    CHECK(steps_.find(tab_name) == steps_.end());
+    steps_[tab_name] = InitializeStep(i);
+  }
+}
+
+const QRect TutorialOverlay::AbsoluteGeometry(QWidget* widget) const {
+  if (widget->parentWidget() == nullptr || widget->parentWidget() == parentWidget()) {
+    return widget->geometry();
+  }
+
+  QWidget* parent_widget = widget->parentWidget();
+  QRect result = widget->geometry();
+  QRect parent_geo = AbsoluteGeometry(parent_widget);
+
+  auto width = result.width();
+  auto height = result.height();
+
+  result.setX(result.x() + parent_geo.x());
+  result.setY(result.y() + parent_geo.y());
+
+  result.setWidth(width);
+  result.setHeight(height);
+
+  return result;
+}
+
+TutorialOverlay::~TutorialOverlay() {}
+
+void TutorialOverlay::StartActiveStep() {
+  const Step& step = GetActiveStep();
+  ui_->tabWidget->setCurrentIndex(step.tab_index);
+
+  if (step.setup.callback_init) {
+    step.setup.callback_init(this, step.name);
+  }
+
+  ui_->btnPrev->setEnabled(!ActiveStepIsFirstInSection());
+  const Section& section = GetActiveSection();
+  const std::string text = absl::StrFormat("%s: %d/%d", section.title,
+                                           (section.active_step - section.step_names.begin()) + 1,
+                                           section.step_names.size());
+  ui_->labelStep->setText(text.c_str());
+
+  if (step.setup.anchor_widget) {
+    step.setup.anchor_widget->installEventFilter(this);
+  }
+  showMaximized();
+  UpdateOverlayLayout();
+}
+
+bool TutorialOverlay::EndActiveStep(bool verify) {
+  if (!HasActiveStep()) {
```

Comment:
Maybe the name "verify" is a bit misleading, since it's more meant to check if anything is preventing us from ending the current step. The logic here is "if there is no active step, nothing prevents us from going further".

But I've realized that `EndActiveStep(true)` is only used in one place anyways, so I've removed the parameter and call `VerifyActiveStep` now explicitly in this one location. I believe that makes it more readable, thanks!